src/Symfony/Component/Translation/DataCollectorTranslator.php

```diff
@@ -47,9 +47,17 @@ public function __construct(TranslatorInterface $translator)
     /**
      * {@inheritdoc}
      */
-    public function trans(?string $id, array $parameters = [], string $domain = null, string $locale = null)
+    public function trans($id, array $parameters = [], string $domain = null, string $locale = null)
     {
-        $trans = $this->translator->trans($id = (string) $id, $parameters, $domain, $locale);
+        if ($id instanceof Translatable) {
+            $parameters += $id->getParameters();
+            $domain = $id->getDomain();
+            $id = $id->getMessageId();
+        }
+
+        $id = (string) $id;
```

Comment:
I would not support casting objects (in any case, that looks like a feature not related to the one introduced by this PR).

Reply:
By having the `$id` strict typed to `?string` before, this would have coerced the object to a string (assuming you don't have `strict_types=1`). Since that was removed, this restores the previous functionality.

Just to clarify, when you say `in any case`, are you referring to all the casting in this PR, or as a phrase?

Reply:
let's preserve the casting in the else case for BC

```
if ($id instanceof Translatable) {
    // ...
    $id = $id->getMessageId();
} else {
    $id = (string) $id;
}
```

Reply:
No longer needed.